app/services/idv/steps/in_person/address_step.rb

```diff
@@ -18,8 +18,11 @@ def analytics_submitted_event
 
         def call
           Idv::InPerson::AddressForm::ATTRIBUTES.each do |attr|
-            next if attr == :same_address_as_id && capture_secondary_id_enabled?
-            flow_session[:pii_from_user][attr] = flow_params[attr]
+            if capture_secondary_id_enabled? && updating_address? && attr == :same_address_as_id
+              flow_session[:pii_from_user][:same_address_as_id] = 'false'
+            else
```

Comment:
What is the value of `flow_session[:pii_from_user][:same_address_as_id]` set to when a user who had it set to `'false'` first lands on the Verify Info page?

Reply:
It is still set to `"false"`.

Does that make you think this line should be changed to do nothing if already `'false'`, (or add another conditional in the `if` statement?

Reply:
Since `Idv::InPerson::AddressForm::ATTRIBUTES` contains `:same_address_as_id` and this field is not submitted by the residential address form during the DAV flow, I suspected that this conditional would reach the `else` case and override the `flow_session[:pii_from_user][:same_address_as_id]` with `nil`.

Reply:
I recommend refactoring the conditional out of the loop and excluding `:same_address_as_id` from iteration when DAV is enabled, e.g.:

```ruby
attrs = Idv::InPerson::AddressForm::ATTRIBUTES

if capture_secondary_id_enabled?
  attrs = attrs.difference([:same_address_as_id])
  flow_session[:pii_from_user][:same_address_as_id] = 'false' if updating_address?
end

attrs.each do |attr|
  flow_session[:pii_from_user][attr] = flow_params[attr]
end
```

Reply:
@NavaTim I think I misinterpreted your original comment. 

When landing on the `Verify Info` page --

If the only code change you make from `main` is to remove the following line from `AddressStep#call`:
 `next if attr == :same_address_as_id && capture_secondary_id_enabled?`

then you're right, the value of `flow_session[:pii_from_user][:same_address_as_id]` on the Verify Info page is `nil`.

The code as it currently stands in this branch returns `'false'`.

I agree that the conditional outside of loop is much cleaner, and thank you for the sample code! I tried doing something similar earlier, and struggled with making it work, but this is so close to working! I think due to my misinterpretation of your first question, I just need one change to your code - I have to move the `if updating_address?` to the outer `if` statement (so `if capture_secondary_id_enabled? && updating_address?`).

Without this, `address_step_spec.rb:79` fails, due to the missing  `same_address_as_id ` `attr` removed by the loop above.

But, with the small change of moving the `updating_address?` check, the (current) tests pass, so I think it's a good change to make.

(Also, TIL I learned `#difference`!)

Reply:
@night-jellyfish When DAV is enabled, the State ID page should be setting `:same_address_as_id` initially instead of the Residential Address page. The test failure you mentioned suggests that the test is incorrect.

Reply:
@NavaTim I _think_ the test is correct because it's checking the contents of the flow session in just the address step, so the state id step has not happened and cannot pass in the value. I think you wouldn't want the address step to be able to set that value in the DAV flow.

Reply:
@JackRyan1989:
> the state id step has not happened and cannot pass in the value

The state ID step will always happen before the address step, without exception. The user will never land on the residential address page without the value of `:same_address_as_id` being set during the DAV flow ([see design](https://figma.com/files/1233938721735543664/project/53695651/Approved-User-Flows-(Internal)?fuid=1119701784920300688)), so a test without that value set is unrealistic.

@night-jellyfish The test referred to on `address_step_spec.rb:79` should be updated to test scenarios with the session's `:same_address_as_id` value set to `'true'` and `'false'` prior to executing the step.

Reply:
Also note that the `same_address_as_id` field [is completely excluded from the address form](https://github.com/18F/identity-idp/blob/c4d3dfd31e2b379c93976cf70b4e43e5199d8c73/app/views/idv/in_person/address.html.erb#LL94C1-L94C1) when DAV is enabled.

Reply:
@NavaTim Thank you for that extra context. With that in mind, do you expect that if:

- _not_ updating the address, and
- DAV enabled

`same_address_as_id` should be `nil`?

I've updated the code back to what you suggested and written tests for the cases as suggested in another comment. I expected that if the above two conditions were true, `:same_address_as_id` would be its initial value, but maybe I'm missing something.

Reply:
@night-jellyfish If the user does not update the residential address and set `same_address_as_id` to true on the State ID page, then it should remain as `'true'`. If the user later updates the residential address, then `same_address_as_id` should be set to `'false'`.

If the user does not update the residential address and set `same_address_as_id` to false on the State ID page, then it should remain `'false'`. If the user then updates the residential address, then `same_address_as_id` should still remain `'false'`.

There is no practical state in which `same_address_as_id` will be `nil` while the user is on the residential address page.

Reply:
@NavaTim That's what I thought, thanks!

I found the `nil` was due to my test setup. But I also found I had to delete this line from your above suggestion: `attrs = attrs.difference([:same_address_as_id])`.

After fixing the test setup and removing that line, I believe I have fixed the issue. But before I push up a patch I wanted to see if there's any reason you might feel that line should be there, and perhaps I'm breaking it in other ways.

Reply:
@night-jellyfish The residential address page in the DAV flow should only ever set `same_address_as_id` to `'false'`, if it sets the value at all. Otherwise it needs to leave the value as-is.

Reply:
Thanks, @NavaTim ! After discussing with @tomas-nava I think we have a good fix in place. I was wrong about needing to delete the line I referenced above. Hopefully tomorrow I'll be able to push up the fix and new tests for review.

Reply:
The implementation has been updated as discussed @NavaTim. Please take a look.